http-gateways/DNSSEC_EXPORT_DNSLINK.md

```diff
@@ -0,0 +1,313 @@
+
+[^1] A longer match in terms of number of subdomains/domains.
+
+```mermaid
+stateDiagram-v2
+    INITIAL --> ENTERING: initial_key_tag
+    ENTERING --> LEAVING
+    LEAVING --> ENTERING: DS
+    LEAVING --> LEAVING: CNAME || DNAME
+    LEAVING --> FAIL
+    ENTERING --> FAIL
+    LEAVING --> ACCEPT: TXT
+```
+
+### CAR File Validation
+This can be done in parallel with the DNSSEC proof validation. We again describe the procedure in terms of a finite state machine. Note that the CAR format supports any hash algorithm that can be encoded in the multihash format. For version 1, we will only support a few common algorithms (e.g., SHA256)—this should not cause any compatibility issues, since the gateway is requesting the CAR file (and thus declaring what hash algorithm the CAR should use).
+
+
+Validation begins in the INITIAL state. Upon reading the CAR header, we extract the CIDs and payloads of each block in the CAR (stored in a list), then enter the VALIDATE\_HASH state. In VALIDATE\_HASH, we check that the client has support for the hash formats and codecs of the current block (and fail if not). We then hash the block payload and check that it matches the CID. If they do not match, validation fails. If they do match, we move to the next block in the list and reenter the VALIDATE\_HASH state. If there are no more blocks, transition to ACCEPT.
+
+
+```mermaid
+stateDiagram-v2
+    INITIAL --> VALIDATE_HASH: Parse blocks, [(CID, payload)]
+    INITIAL --> FAIL: Cannot parse CAR
+    VALIDATE_HASH --> VALIDATE_HASH: CID == hash(payload)
+    VALIDATE_HASH --> FAIL: CID != hash(payload)
+    VALIDATE_HASH --> ACCEPT
+```
+For the client implementation, we will prefer to use an available library. Both Go and JS libraries are available (i.e., go-car and js-car). If we opt for the browser extension design, the JS library is probably the best option.
+
+### Alternative Design
+For each request, the gateway would respond with an additional header of the form X-Ipfs-Name-Proof: <CID>, where <CID> is a pointer to an IPFS file containing the DNSSEC proof. A client could then retrieve this proof directly from IPFS if they want it. However, we decided against this approach for several reasons. The main reason is that it is never more advantageous for the gateway to retrieve the DNSSEC proof from IPFS than it would be to get it from a local cache. The only time this would be useful is if a gateway could use DNSSEC proofs found by other gateways. This is not feasible because the DNSSEC proof would be stored according to its hash (in IPFS), and the hash required can only be known by already collecting the DNSSEC proof—therefore it would be pointless for a gateway to request content that, by definition, it already has. An additional problem is that DNSSEC proofs have TTLs, and thus will not usually be valid for more than an hour. This limits the time that the content would be useful anyway.
```

Comment:
having the resolution in two steps require the gateway to keep some state, which I don't think is a good requirement. Ideally, gateway should be stateless. In the variant you propose above, a stateless gateway could be consider dishonest as it might not necessarily have the same view of DNS in the two requests.

For a two requests protocol to not require the gateway to have state, a client should first request the DNSSEC proof, and then request the resolved CID/CAR.

Reply:
In addition, the gateway doesn't need to request the proofs from the resolver unless it receives the header `"Accept: application/vnd.dnslink.dnssec; version=1"`. So a client who's only interested in the CAR can opt to receive only that.

Reply:
I agree on two points: the gateway should be stateless, and the client should be able to decide when the DNSLink proof is returned.

What I'm not comfortable with is limiting clients, forcing them to always fetch data when all they want is DNSLink proofs.  Having 2-in-1 response type is fine (sounds like you have use cases in mind for this?), but there should also be a way to request proof without the data, allowing clients to do the data retrieval by other means.

> For a two requests protocol to not require the gateway to have state, a client should first request the DNSSEC proof, and then request the resolved CID/CAR.

This is exactly that I had in mind, and it follows the patterns of how the web works.

You need to resolve DNSLink once, and in my mind DNSSEC proof would be part of that.
After that, you don't need it until TTL runs out.